File: MENTORING3/Assets/Flocking/Scripts/Boid.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

//Doesn't have a Prefab yet.
public class Boid : MonoBehaviour
{
    private Vector3 position = new();
    private Quaternion rotation = Quaternion.identity;
    private Vector3 velocity;
    private Vector3 acceleration;
    private float maxForce; //Controls, how fast they align.
    private float maxSpeed;
    private int boidsView; //perseption
    private GameObject boidRef;

    // Start is called before the first frame update
    public void Start()
    {
        velocity = MathUtilities.Random(new Vector3(-1.1f, 1.1f, 1.1f), new Vector3(-1.1f, 1.1f, 1.1f));
        acceleration = MathUtilities.Random(new Vector3(-1.1f, 1.1f, 1.1f), new Vector3(-1.1f, 1.1f, 1.1f));
    }

    public void Setup(float initialMaxForce, float initialMaxSpeed, int initialBoidsView, GameObject boid)
    {
        rotation.eulerAngles = MathUtilities.Random(new Vector3(-5f, 5f, 5f), new Vector3(-5f, 5f, 5f));
        maxForce = initialMaxForce;
        maxSpeed = initialMaxSpeed;
        boidsView = initialBoidsView;
        boidRef = boid;
        boidRef.transform.Translate(position);

        //Debugging!!!
        //Component[] components = boidRef.GetComponents(typeof(Component));
        //foreach (Component component in components)
        //{
        //    Debug.Log("Component: "+component.ToString());
        //}
    }

    //Donut world
    public void Edges()
    {
        //Debug.Log("entred Edges");
        if (position.x > 5)
        {
            position.x = 0;
        }
        else if (position.x < 0)
        {
            position.x = 5;
        }

        if (position.y > 5)
        {
            position.y = 0;
        }
        else if (position.y < 0)
        {
            position.y = 5;
        }

        if (position.y > 5)
        {
            position.y = 0;
        }
        else if (position.y < 0)
        {
            position.y = 5;
        }
    }

    Vector3 Align(Boid[] flock)
    {
        //Debug.Log("entred Align");
        Vector3 steering = new(); //Vector (force) to correct the direction
        int total = 0;

        foreach (Boid otherBoid in flock)
        {
            float distance = Vector3.Distance(position, otherBoid.position);
            if (otherBoid != this && distance <= boidsView)
            {
                steering += otherBoid.velocity;
                total++;
            }
        }

        if (total > 0)
        {
            steering /= total;
            steering = Vector3.ClampMagnitude(steering, maxSpeed);
            steering -= velocity;
            steering = Vector3.ClampMagnitude(steering, maxForce);
        }
        return steering;
    }

    Vector3 Cohesion(Boid[] flock)
    {
        //Debug.Log("entred Cohesion");
        Vector3 steering = new();
        int total = 0;

        foreach (Boid otherBoid in flock)
        {
            float distance = Vector3.Distance(position, otherBoid.position);
            if (otherBoid != this && distance <= boidsView)
            {
                steering += otherBoid.position;
                total++;
            }
        }

        if (total > 0)
        {
            steering /= total;
            steering -= position;
            steering = Vector3.ClampMagnitude(steering, maxSpeed);
            steering -= velocity;
            steering = Vector3.ClampMagnitude(steering, maxForce);
        }
        return steering;
    }

    Vector3 Separation(Boid[] flock)
    {
        //Debug.Log("entred Separation");
        Vector3 steering = new();
        int total = 0;

        foreach (Boid otherBoid in flock)
        {
            float distance = Vector3.Distance(position, otherBoid.position);
            if (otherBoid != this && distance <= boidsView)
            {
                Vector3 diff = position - otherBoid.position;
                diff /= distance * distance;
                steering += diff;
                total++;
            }
        }

        if (total > 0)
        {
            steering /= total;
            steering = Vector3.ClampMagnitude(steering, maxSpeed);
            steering -= velocity;
            steering = Vector3.ClampMagnitude(steering, maxForce);
        }
        return steering;
    }

    //Apply rules
    public void Flock(Boid[] flock)
    {
        //Debug.Log("entred Flock");
        Vector3 alignment = Align(flock);
        Vector3 cohesion = Cohesion(flock);
        Vector3 separation = Separation(flock);

        acceleration += alignment;
        acceleration += cohesion;
        acceleration += separation;
    }

    public void UpdateBoid()
    {
        //Debug.Log("entred UpdateBoid");
        boidRef.transform.Translate(position);
        position += velocity;
        velocity += acceleration;
        velocity = Vector3.ClampMagnitude(velocity, maxSpeed);
        acceleration *= 0; //Reset vector
    }
}

File: p5js/Schwarm/p5project swarm/boid.js
class Boid {
  constructor() {
    this.position = createVector(random(width), random(height));
    this.velocity = p5.Vector.random2D();
    this.velocity.setMag(random(2, 4));
    this.acceleration = createVector();
    this.maxForce = initialmaxForce; //Controls, how fast they align.
    this.maxSpeed = initialSpeed;
    this.boidsView = initialBoidsView; //perseption
  }

  //donut world
  edges(){
    if(this.position.x > width) {
        this.position.x = 0;
    } else if(this.position.x < 0) {
        this.position.x = width;
    }
    if(this.position.y > width) {
        this.position.y = 0;
    } else if(this.position.y < 0) {
        this.position.y = width;
    }
  }

  align(boids) {
    let steering = createVector(); //Vector (force) to correct the direction
    let total = 0;
    for (let other of boids) {
      let distance = dist(
        this.position.x,
        this.position.y,
        other.position.x,
        other.position.y,
      );

      //Just look at boids in view
      if (other != this && distance <= this.boidsView) {
        steering.add(other.velocity);
        total++;
      }
    }
    if (total > 0) {
      steering.div(total);
      steering.setMag(this.maxSpeed);
      steering.sub(this.velocity);
      steering.limit(this.maxForce); //Limits magnitude
    }
    return steering;
  }

  cohesion(boids) {
    let steering = createVector();
    let total = 0;
    for (let other of boids) {
      let distance = dist(
        this.position.x,
        this.position.y,
        other.position.x,
        other.position.y,
      );

      if (other != this && distance <= this.boidsView) {
        steering.add(other.position);
        total++;
      }
    }
    if (total > 0) {
      steering.div(total);
      steering.sub(this.position);
      steering.setMag(this.maxSpeed);
      steering.sub(this.velocity);
      steering.limit(this.maxForce);
    }
    return steering;
  }

  separation(boids) {
    let steering = createVector();
    let total = 0;
    for (let other of boids) {
      let distance = dist(
        this.position.x,
        this.position.y,
        other.position.x,
        other.position.y,
      );

      if (other != this && distance <= this.boidsView*2) {
        let diff = p5.Vector.sub(this.position, other.position);
        diff.div(distance * distance);
        steering.add(diff);
        total++;
      }
    } 
    if (total > 0) {
      steering.div(total);
      steering.setMag(this.maxSpeed);
      steering.sub(this.velocity);
      steering.limit(this.maxForce);
    }
    return steering;
  }

  //Apply rules
  flock(boids) {
    let alignment = this.align(boids);
    let cohesion = this.cohesion(boids);
    let separation = this.separation(boids);
    
    this.acceleration.add(alignment);
    this.acceleration.add(cohesion);
    this.acceleration.add(separation);

    //Update with slider's value
    this.maxForce = maxForceSlider.value();
    this.maxSpeed = maxSpeedSlider.value();
    this.boidsView = boidsViewSlider.value();
    alignment.mult(alignSlider.value());
    cohesion.mult(cohesionSlider.value());
    separation.mult(separationSlider.value());
  }

  updateBoids() {
    this.position.add(this.velocity);
    this.velocity.add(this.acceleration);
    this.velocity.limit(this.maxSpeed);
    this.acceleration.mult(0); //Reset vector
  }

  showBoids() {
    strokeWeight(8);
    stroke(255);
    point(this.position.x, this.position.y);
  }
}
